File: src/Ccmc.Core/Services/ProjectFilter.cs
using Ccmc.Core.Models;

namespace Ccmc.Core.Services;

/// <summary>A snapshot of the facts a SavedFilter tests. The VM fills this from its rows.</summary>
public readonly record struct ProjectFacts(
    string Path, bool HasGit, bool HasClaudeMd, bool IsRunning, bool IsPinned);

/// <summary>Pure evaluation of a SavedFilter against ProjectFacts. AND semantics; unset = pass.</summary>
public static class ProjectFilter
{
    public static bool Matches(SavedFilter filter, ProjectFacts facts)
    {
        ArgumentNullException.ThrowIfNull(filter);
        if (!string.IsNullOrWhiteSpace(filter.PathContains) &&
            facts.Path.IndexOf(filter.PathContains, StringComparison.OrdinalIgnoreCase) < 0)
            return false;
        if (filter.RequireGit && !facts.HasGit) return false;
        if (filter.RequireClaudeMd && !facts.HasClaudeMd) return false;
        if (filter.RequireRunning && !facts.IsRunning) return false;
        if (filter.RequirePinned && !facts.IsPinned) return false;
        return true;
    }
}

File: electron/src/core/projects/projectFilter.ts
import type { SavedFilter } from '../models'

/** A snapshot of the facts a SavedFilter tests. The VM fills this from its rows. */
export interface ProjectFacts {
  path: string
  hasGit: boolean
  hasClaudeMd: boolean
  isRunning: boolean
  isPinned: boolean
}

/** Pure evaluation of a SavedFilter against ProjectFacts. AND semantics; unset = pass. */
export function filterMatches(filter: SavedFilter, facts: ProjectFacts): boolean {
  if (
    filter.pathContains &&
    filter.pathContains.trim() !== '' &&
    !facts.path.toLowerCase().includes(filter.pathContains.toLowerCase())
  ) {
    return false
  }
  if (filter.requireGit && !facts.hasGit) return false
  if (filter.requireClaudeMd && !facts.hasClaudeMd) return false
  if (filter.requireRunning && !facts.isRunning) return false
  if (filter.requirePinned && !facts.isPinned) return false
  return true
}
